fleetspeak/src/server/stats.go

```diff
@@ -142,50 +221,66 @@ var (
 }
 
 func (s PrometheusStatsCollector) MessageProcessed(start, end time.Time, service, messageType string) {
-	messagesProcessed.Inc()
+	messagesProcessed.WithLabelValues(messageType).Observe(end.Sub(start).Seconds())
 }
 
 func (s PrometheusStatsCollector) MessageErrored(start, end time.Time, service, messageType string, isTemp bool) {
-	messagesErrored.Inc()
-	if isTemp {
-		messagesErroredTemp.Inc()
-	} else {
-		messagesErroredPerm.Inc()
-	}
+	messagesErrored.WithLabelValues(messageType, strconv.FormatBool(isTemp)).Observe(end.Sub(start).Seconds())
 }
 
 func (s PrometheusStatsCollector) MessageDropped(service, messageType string) {
-	messagesDropped.Inc()
+	messagesDropped.WithLabelValues(service, messageType).Inc()
 }
 
 func (s PrometheusStatsCollector) ClientPoll(info stats.PollInfo) {
-	clientPolls.Inc()
+	clientPolls.WithLabelValues(strconv.Itoa(info.Status), info.Type.String(), strconv.FormatBool(info.CacheHit)).Inc()
+	clientPollsOpTime.WithLabelValues(strconv.Itoa(info.Status), info.Type.String(), strconv.FormatBool(info.CacheHit)).Observe(info.End.Sub(info.Start).Seconds())
+	clientPollsReadTime.WithLabelValues(strconv.Itoa(info.Status), info.Type.String(), strconv.FormatBool(info.CacheHit)).Observe(info.ReadTime.Seconds())
+	clientPollsWriteTime.WithLabelValues(strconv.Itoa(info.Status), info.Type.String(), strconv.FormatBool(info.CacheHit)).Observe(info.WriteTime.Seconds())
+	clientPollsReadBytes.WithLabelValues(strconv.Itoa(info.Status), info.Type.String(), strconv.FormatBool(info.CacheHit)).Observe(float64(info.ReadBytes / 1000000))
+	clientPollsWriteBytes.WithLabelValues(strconv.Itoa(info.Status), info.Type.String(), strconv.FormatBool(info.CacheHit)).Observe(float64(info.WriteBytes / 1000000))
 }
 
 func (s PrometheusStatsCollector) DatastoreOperation(start, end time.Time, operation string, result error) {
-	datastoreOperationsCompleted.Inc()
+	datastoreOperationsCompleted.WithLabelValues(operation, strconv.FormatBool(result != nil)).Observe(end.Sub(start).Seconds())
+}
+
+func getClientDataLabelsConcatenated(cd *db.ClientData) string {
+	var clientDataLabels []string
+	for _, labelStruct := range cd.Labels {
+		clientDataLabels = append(clientDataLabels, labelStruct.GetLabel())
+	}
+	sort.Strings(clientDataLabels)
+	return strings.Join(clientDataLabels[:], ",")
 }
 
 func (s PrometheusStatsCollector) ResourceUsageDataReceived(cd *db.ClientData, rud mpb.ResourceUsageData, v *fspb.ValidationInfo) {
-	resourcesUsageDataReceived.Inc()
+	clientDataLabels := getClientDataLabelsConcatenated(cd)
+
+	// Counter
+	resourcesUsageDataReceivedCount.WithLabelValues(clientDataLabels, strconv.FormatBool(cd.Blacklisted), rud.Scope, rud.Version).Inc()
+
+	// Historgrams
+	resourcesUsageDataReceivedByMeanUserCPURate.WithLabelValues(clientDataLabels, strconv.FormatBool(cd.Blacklisted), rud.Scope, rud.Version).Observe(rud.ResourceUsage.GetMeanUserCpuRate())
```

Comment:
Same issue as above - please put labels values into variables. The code will be easier to read and faster to run. Right now strconv.FormatBool is called a lot and the code feels cluttered.

Reply:
I agree that the code will run slower in the case of the label `blacklisted` if we keep using `strconv.FormatBool`. But, out of curiosity, in the cases of `scope` and `version`, wouldn't the code run in the same speed, given that the `rud` is passed by value?
Regardless, I will put all labels values into variables for consistency and readability.

Reply:
Yes, in case of `scope` and `version` the performance is not an issue. To be honest, the performance likely wouldn't be an issue for other labels as well, as the practical overhead would likely be minimal. However, not calling the same function multiple times with the same set of arguments makes the code more readable and easier to follow. The new versions of the code looks very good.